test/concurrent.cc

```diff
@@ -0,0 +1,717 @@
+                     } else {
+                       static_cast<K*>(data.k)->Stop();
+                     }
+                   });
+                 }));
+           })
+        | Collect<std::vector<std::string>>();
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  ASSERT_EQ(2, callbacks.size());
+
+  EXPECT_EQ(
+      std::future_status::timeout,
+      future.wait_for(std::chrono::seconds(0)));
+
+  for (auto& callback : callbacks) {
+    callback();
+  }
+
+  EXPECT_THROW(future.get(), eventuals::StoppedException);
+}
+
+TEST(ConcurrentTest, Fail) {
+  std::deque<Callback<>> callbacks;
+
+  auto e = [&]() {
+    return Iterate({1, 2})
+        | Concurrent([&]() {
+             struct Data {
+               void* k;
+               int i;
+             };
+             return Map(Eventual<std::string>(
+                 [&, data = Data()](auto& k, int i) mutable {
+                   using K = std::decay_t<decltype(k)>;
+                   data.k = &k;
+                   data.i = i;
+                   callbacks.emplace_back([&data]() {
+                     if (data.i == 1) {
+                       static_cast<K*>(data.k)->Start(std::to_string(data.i));
+                     } else {
+                       static_cast<K*>(data.k)->Fail("error");
+                     }
+                   });
+                 }));
+           })
+        | Collect<std::vector<std::string>>();
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  ASSERT_EQ(2, callbacks.size());
+
+  EXPECT_EQ(
+      std::future_status::timeout,
+      future.wait_for(std::chrono::seconds(0)));
+
+  for (auto& callback : callbacks) {
+    callback();
+  }
+
+  EXPECT_THROW(future.get(), std::exception_ptr);
+}
+
+
+TEST(ConcurrentTest, FailOrStop) {
+  std::deque<Callback<>> callbacks;
+
+  auto e = [&]() {
+    return Iterate({1, 2})
+        | Concurrent([&]() {
+             struct Data {
+               void* k;
+               int i;
+             };
+             return Map(Eventual<std::string>(
+                 [&, data = Data()](auto& k, int i) mutable {
+                   using K = std::decay_t<decltype(k)>;
+                   data.k = &k;
+                   data.i = i;
+                   callbacks.emplace_back([&data]() {
+                     if (data.i == 1) {
+                       static_cast<K*>(data.k)->Stop();
+                     } else {
+                       static_cast<K*>(data.k)->Fail("error");
+                     }
+                   });
+                 }));
+           })
+        | Collect<std::vector<std::string>>();
+  };
+
+  auto [future, k] = Terminate(e());
+
+  k.Start();
+
+  ASSERT_EQ(2, callbacks.size());
+
+  EXPECT_EQ(
+      std::future_status::timeout,
+      future.wait_for(std::chrono::seconds(0)));
+
+  for (auto& callback : callbacks) {
+    callback();
+  }
+
+  EXPECT_ANY_THROW(future.get());
+}
+
+TEST(ConcurrentTest, InterruptSuccess) {
```

Comment:
I am not entirely sure what this test tests. Why there's an interrupt and why the whole eventual succeeds even after triggering the interrupt? Please add a comment!